libafl_frida/src/alloc.rs

```diff
@@ -0,0 +1,408 @@
+    current_mapping_addr: usize,
+}
+
+macro_rules! map_to_shadow {
+    ($self:expr, $address:expr) => {
+        (($address >> 3) + $self.shadow_offset) & ((1 << ($self.shadow_bit + 1)) - 1)
+    };
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub(crate) struct AllocationMetadata {
+    pub address: usize,
+    pub size: usize,
+    pub actual_size: usize,
+    pub allocation_site_backtrace: Option<Backtrace>,
+    pub release_site_backtrace: Option<Backtrace>,
+    pub freed: bool,
+    pub is_malloc_zero: bool,
+}
+
+impl Allocator {
+    pub fn new(options: FridaOptions) -> Self {
+        let ret = unsafe { sysconf(_SC_PAGESIZE) };
+        if ret < 0 {
+            panic!("Failed to read pagesize {:?}", io::Error::last_os_error());
+        }
+        #[allow(clippy::cast_sign_loss)]
+        let page_size = ret as usize;
+        // probe to find a usable shadow bit:
+        let mut shadow_bit: usize = 0;
+        for try_shadow_bit in &[46usize, 36usize] {
+            let addr: usize = 1 << try_shadow_bit;
+            if unsafe {
+                mmap(
+                    addr as *mut c_void,
+                    page_size,
+                    ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+                    MapFlags::MAP_PRIVATE
+                        | MapFlags::MAP_ANONYMOUS
+                        | MapFlags::MAP_FIXED
+                        | MapFlags::MAP_NORESERVE,
+                    -1,
+                    0,
+                )
+            }
+            .is_ok()
+            {
+                shadow_bit = *try_shadow_bit;
+                break;
+            }
+        }
+        assert!(shadow_bit != 0);
+
+        // attempt to pre-map the entire shadow-memory space
+        let addr: usize = 1 << shadow_bit;
+        let pre_allocated_shadow = unsafe {
+            mmap(
+                addr as *mut c_void,
+                addr + addr,
+                ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+                MapFlags::MAP_ANONYMOUS
+                    | MapFlags::MAP_FIXED
+                    | MapFlags::MAP_PRIVATE
+                    | MapFlags::MAP_NORESERVE,
+                -1,
+                0,
+            )
+        }
+        .is_ok();
+
+        Self {
+            options,
+            page_size,
+            pre_allocated_shadow,
+            shadow_offset: 1 << shadow_bit,
+            shadow_bit,
+            allocations: HashMap::new(),
+            shadow_pages: RangeSet::new(),
+            allocation_queue: HashMap::new(),
+            largest_allocation: 0,
+            base_mapping_addr: addr + addr + addr,
+            current_mapping_addr: addr + addr + addr,
+        }
+    }
+
+    /// Retreive the shadow bit used by this allocator.
+    pub fn shadow_bit(&self) -> u32 {
+        self.shadow_bit as u32
+    }
+
+    #[inline]
+    fn round_up_to_page(&self, size: usize) -> usize {
+        ((size + self.page_size) / self.page_size) * self.page_size
```

Comment:
Not sure if related, but there was an align function in llmp, maybe you can reuse that or move to bolts?

Reply:
Hm I guess different context, don't bother :)

Reply:
Hm I guess different context, don't bother :)

Reply:
Not bothering.